internal/app/server/impl.go

```diff
@@ -0,0 +1,374 @@
+	ctx context.Context,
+	req *connect.Request[v1alpha1.UnaryRequest],
+) (*connect.Response[v1alpha1.UnaryResponse], error) {
+	msgAsAny, err := asAny(req.Msg)
+	if err != nil {
+		return nil, err
+	}
+	payload, err := parseUnaryResponseDefinition(req.Msg.ResponseDefinition, req.Header(), []*anypb.Any{msgAsAny})
+	if err != nil {
+		return nil, err
+	}
+
+	resp := connect.NewResponse(&v1alpha1.UnaryResponse{
+		Payload: payload,
+	})
+
+	if req.Msg.ResponseDefinition != nil {
+		setResponseHeadersAndTrailers(req.Msg.ResponseDefinition, resp)
+	}
+
+	return resp, nil
+}
+
+func (s *conformanceServer) ClientStream(
+	ctx context.Context,
+	stream *connect.ClientStream[v1alpha1.ClientStreamRequest],
+) (*connect.Response[v1alpha1.ClientStreamResponse], error) {
+	var responseDefinition *v1alpha1.UnaryResponseDefinition
+	firstRecv := true
+	var reqs []*anypb.Any
+	for stream.Receive() {
+		if err := ctx.Err(); err != nil {
+			return nil, err
+		}
+		msg := stream.Msg()
+		// If this is the first message received on the stream, save off the response definition we need to send
+		if firstRecv {
+			responseDefinition = msg.ResponseDefinition
+			firstRecv = false
+		}
+		// Record all the requests received
+		msgAsAny, err := asAny(msg)
+		if err != nil {
+			return nil, err
+		}
+		reqs = append(reqs, msgAsAny)
+	}
+	if err := stream.Err(); err != nil {
+		return nil, err
+	}
+
+	payload, err := parseUnaryResponseDefinition(responseDefinition, stream.RequestHeader(), reqs)
+
+	resp := connect.NewResponse(&v1alpha1.ClientStreamResponse{
+		Payload: payload,
+	})
+
+	if responseDefinition != nil {
+		setResponseHeadersAndTrailers(responseDefinition, resp)
+	}
+
+	return resp, err
+}
+
+func (s *conformanceServer) ServerStream(
+	ctx context.Context,
+	req *connect.Request[v1alpha1.ServerStreamRequest],
+	stream *connect.ServerStream[v1alpha1.ServerStreamResponse],
+) error {
+	responseDefinition := req.Msg.ResponseDefinition
+	if responseDefinition != nil {
+		// Set all requested response headers on the response
+		for _, header := range responseDefinition.ResponseHeaders {
+			for _, val := range header.Value {
+				stream.ResponseHeader().Add(header.Name, val)
+			}
+		}
+		// Set all requested response trailers on the response
+		for _, trailer := range responseDefinition.ResponseTrailers {
+			for _, val := range trailer.Value {
+				stream.ResponseTrailer().Add(trailer.Name, val)
+			}
+		}
+	}
+
+	// Convert the request to an Any so that it can be recorded in the payload
+	msgAsAny, err := asAny(req.Msg)
+	if err != nil {
+		return err
+	}
+	payload := initPayload(req.Header(), []*anypb.Any{msgAsAny})
+
+	for _, data := range responseDefinition.ResponseData {
+		payload.Data = data
+
+		resp := &v1alpha1.ServerStreamResponse{
+			Payload: payload,
+		}
+
+		if err := sendOnStream[v1alpha1.ServerStreamResponse](
+			ctx,
+			stream,
+			resp,
+			responseDefinition.ResponseDelayMs,
+		); err != nil {
+			return err
+		}
+
```

Comment:
I think you want to clear `payload.RequestInfo` here, so we only echo back the request in the first response.

I was reading back at the proto, and it's unclear what to do when `ResponseData` is empty. The comment says to send back no headers, trailers, or data in each response, but I guess I didn't read that closely enough during review to call that out -- "in each response" doesn't make sense, since we don't know how many to send back.

I definitely think it's useful to test a _zero_-sized response stream, which implies that if `ResponseData` is empty/unset, the server sends back no messages. That is what is actually implemented here, and I think that is the most sensible behavior. The one downside is that we have no way of echoing back the actual request data, to make sure it survives a round-trip. But I guess that is adequately tested in test cases where stream has at least one response message 🤷.